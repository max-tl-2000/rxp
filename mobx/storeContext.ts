/*
 * Copyright (c) 2022 Reva Technology Inc., all rights reserved.
 * Unauthorized copying of this file, via any medium is strictly prohibited
 * Licensed under the Elastic License 2.0; you may not use this file except
 * in compliance with the Elastic License 2.0.
 */
import { createContext } from 'react';
import { configure, reaction } from 'mobx';
import { Auth } from './stores/auth';
import { UserSettings } from './stores/userSettings';
import { mediator, makeRequest } from '../network/helpers';
import { syncStorage } from '../helpers/sync-storage/sync-storage';
import { User } from './stores/authTypes';
import { isWeb } from '../config';
import { Stores } from './stores/Stores';
import { initLogger, disableLogger } from '../helpers/logger';

configure({ enforceActions: 'observed' });

type OnUserChangeCallback = {
  (user?: User): void;
};

type OnPropertyChangeCallback = {
  (propertyId?: string): void;
};

const onUserChange = (auth: Auth, cb: OnUserChangeCallback) =>
  reaction(
    () => {
      const { user } = auth;
      return { user };
    },
    ({ user }) => cb(user),
  );

const onPropertyIdChange = (userSettings: UserSettings, cb: OnPropertyChangeCallback) =>
  reaction(
    () => {
      const { selectedPropertyId } = userSettings;
      return { selectedPropertyId };
    },
    ({ selectedPropertyId }) => cb(selectedPropertyId),
  );

export const createMobxStores = (): Stores => {
  const stores = new Stores();

  const { auth, settings, userSettings } = stores;

  auth.onBeforeLogout = () => {
    disableLogger();
    mediator().next('user:before-logout');
  };

  onUserChange(auth, async (user?: User) => {
    if (user) {
      syncStorage.fields.userData = user;

      // had to move this here to avoid the error about the logger not being initialized
      initLogger(auth, settings, userSettings, makeRequest);

      mediator().next('user:login', user);

      mediator().next('user:loggedIn', user);
    } else {
      syncStorage.entries.userData.clear();

      // by this time there is no userData anymore
      mediator().next('user:logout');

      // we reset the stores after emitting the event in case they need to access the store for some reason
      stores.resetStores();
    }
  });

  onPropertyIdChange(userSettings, async (propertyId?: string) => {
    if (propertyId) {
      mediator().next('property:change', propertyId);
      stores.resetStores();
    }
  });

  if (isWeb) {
    const onStorageChange = async (e: StorageEvent) => {
      if (e.key === 'userData') {
        await syncStorage.entries.userData.load();
        if (!syncStorage.fields.userData) {
          auth.signOut();
        }
      }
    };

    window.addEventListener('storage', onStorageChange, false);
  }

  return stores;
};

export const MobxStoresContext = createContext<Stores>({} as Stores);
