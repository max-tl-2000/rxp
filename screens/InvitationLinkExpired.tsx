import React from 'react';
import { StyleSheet, TextStyle } from 'react-native';
import { Title, Text } from 'react-native-paper';
import { t } from '../i18n/trans';

import { ContentWrapper } from '../components';
import { AppBranding } from '../components/AppBranding';

interface Styles {
  title: TextStyle;
  message: TextStyle;
}

const createStyles = () => {
  const baseStyles: Styles = {
    title: {
      textAlign: 'center',
      marginBottom: 24,
    },
    message: {
      textAlign: 'center',
      fontSize: 13,
      lineHeight: 20,
      marginBottom: 24,
    },
  };
  return StyleSheet.create({ ...baseStyles });
};

export const InvitationLinkExpired = (): JSX.Element => {
  const styles = createStyles();

  return (
    <ContentWrapper isAuthScreenContainer>
      <AppBranding />
      <Title style={styles.title}>{t('SORRY_INVITATION_HAS_EXPIRED')}</Title>
      <Text style={styles.message}>{t('INVITATION_HAS_EXPIRED_INSTRUCTIONS')}</Text>
    </ContentWrapper>
  );
};
