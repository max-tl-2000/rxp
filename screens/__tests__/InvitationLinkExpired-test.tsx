import * as React from 'react';
import renderer, { act, ReactTestRenderer } from 'react-test-renderer';
import { InvitationLinkExpired } from '../InvitationLinkExpired';
import { MobxStoresContext, createMobxStores } from '../../mobx/storeContext';

jest.mock('../../components/AppBranding', () => {
  const { View } = require('react-native'); // eslint-disable-line
  return {
    AppBranding() {
      return <View />;
    },
  };
});

jest.mock('@react-navigation/stack', () => ({
  useHeaderHeight: () => 62,
}));

describe('InvitationLinkExpired', () => {
  beforeEach(() => {
    jest.useFakeTimers();
  });

  it('renders the InvitationLinkExpired component', async () => {
    let resolve: (value: ReactTestRenderer) => void;

    const renderPromise: Promise<ReactTestRenderer> = new Promise(r => {
      resolve = r;
    });

    await act(async () =>
      resolve(
        renderer.create(
          <MobxStoresContext.Provider value={createMobxStores()}>
            <InvitationLinkExpired />
          </MobxStoresContext.Provider>,
        ),
      ),
    );

    const rdr = await renderPromise;
    expect(rdr.toJSON()).toMatchSnapshot();
  });
});
