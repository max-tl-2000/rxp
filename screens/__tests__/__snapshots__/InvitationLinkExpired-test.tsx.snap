// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InvitationLinkExpired renders the InvitationLinkExpired component 1`] = `
<View
  onLayout={[Function]}
  style={
    Array [
      Object {
        "backgroundColor": "#f6f6f6",
        "flex": 1,
      },
      Object {
        "paddingBottom": 0,
      },
    ]
  }
>
  <RCTScrollView
    contentContainerStyle={
      Object {
        "backgroundColor": "#f6f6f6",
        "justifyContent": "flex-end",
      }
    }
  >
    <View>
      <View
        accessible={false}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "alignItems": "center",
            "backgroundColor": "#f6f6f6",
            "marginTop": 62,
            "paddingHorizontal": 36,
            "width": "100%",
          }
        }
      >
        <View />
        <Text
          style={
            Array [
              Object {
                "color": "#000000",
                "fontFamily": "System",
                "fontWeight": "400",
                "textAlign": "left",
              },
              Array [
                Object {
                  "color": "rgba(0, 0, 0, 0.87)",
                  "fontFamily": "System",
                  "fontWeight": "500",
                  "textAlign": "left",
                  "writingDirection": "ltr",
                },
                Array [
                  Object {
                    "fontSize": 20,
                    "letterSpacing": 0.15,
                    "lineHeight": 30,
                    "marginVertical": 2,
                  },
                  Object {
                    "marginBottom": 24,
                    "textAlign": "center",
                  },
                ],
              ],
            ]
          }
        >
          Sorry!
This invitation has expired.
        </Text>
        <Text
          style={
            Array [
              Object {
                "color": "#000000",
                "fontFamily": "System",
                "fontWeight": "400",
                "textAlign": "left",
              },
              Object {
                "fontSize": 13,
                "lineHeight": 20,
                "marginBottom": 24,
                "textAlign": "center",
              },
            ]
          }
        >
          To register, you need to have a new invitation sent to
you by one of our leasing agents. Please contact
 your agent and ask for an invitation, so you can register a new account.
        </Text>
        <View
          style={
            Object {
              "backgroundColor": "#f6f6f6",
              "flex": 1,
            }
          }
        />
      </View>
    </View>
  </RCTScrollView>
</View>
`;
